### PR TITLE
Add opam files to allow pinning

### PR DESCRIPTION
**Pull request by gretay-js**
Address comments from PR #2207.

Additionally, attached ocaml-4.08.0.opam file should be added to opam repository.

[ocaml-4.08.0.opam.txt](https://github.com/ocaml/ocaml/files/3063157/ocaml-4.08.0.opam.txt)

### CODE REVIEW

File: HACKING.adoc [hunk 1]
```diff
@@ -43,6 +43,14 @@ make tests
...
 opam compiler-conf install
 ----
++
+With opam 2, create a local opam switch with the compiler installed from
+the current source directory:
++
+----
+opam switch create . --empty
+opam install .
+----
```
Comment (Drup) [inline HACKING.adoc:53]: Why advise a local switch for this ?

File: ocaml-variants.opam [hunk 2]
```diff
@@ -0,0 +1,30 @@
+opam-version: "2.0"
+name: "ocaml-variants"
+version: "4.08.0+dev"
```
Comment (Drup) [inline ocaml-variants.opam]: The usual name for ocaml dev channel is "trunk"

File: ocaml-variants.opam [hunk 3]
```diff
@@ -0,0 +1,30 @@
+opam-version: "2.0"
+name: "ocaml-variants"
```
Comment (Drup) [inline ocaml-variants.opam]: You don't need the name here.

File: ocaml-variants.opam [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+flags: compiler
+setenv: CAML_LD_LIBRARY_PATH = "%{lib}%/stublibs"
+build: [
+  ["./configure" "--prefix=%{prefix}%"]
+  [make "-j%{jobs}%" "world"]
+  [make "-j%{jobs}%" "world.opt"]
+]
+install: [make "install"]
+maintainer: "caml-list@inria.fr"
+homepage: "https://ocaml.org"
```
Comment (Drup) [inline ocaml-variants.opam]: Given this is the dev version, I think putting the github page here is more appropriate.

File: ocaml-variants.opam [hunk 5]
```diff
@@ -0,0 +1,30 @@
+opam-version: "2.0"
+name: "ocaml-variants"
+version: "4.08.0+dev"
+synopsis: "OCaml development version"
+depends: [
+  "ocaml" {= "4.08.0" & post}
```
Comment (Drup) [inline ocaml-variants.opam:5]: IIRC, we can put `= version` here, and it'll do the right thing directly.

Comment (gretay-js) [reply ocaml-variants.opam:5]: I have just tried it with `
depends: [
  "ocaml" {= "%{version}%" & post} ...`
It fails when I try to create a local repo from a global system switch that has 4.07.1. 
```
The following dependencies couldn't be met:
  - ocaml-variants -> ocaml >= 4.08.0+trunk
      no matching version
```
Is this the behavior you expect? With the explicit dependency on 4.08.0 & post, opam finds ocaml.opam (currently in the same directory) and installs ocaml-variants successfully.

File: ocaml-variants.opam (hunk 4 above)
Comment (gretay-js) [reply ocaml-variants.opam]: Fixed this and the other comments above.

File: ocaml-variants.opam (hunk 5 above)
Comment (Drup) [reply ocaml-variants.opam:5]: Ah, wait, the version is `4.08.0+trunk`, but we want `4.08.0`. Nevermind, disregard that remark. :)

**Review by Drup — APPROVED** [review]
Looks good!

File: HACKING.adoc (hunk 1 above)
Comment (gasche) [reply HACKING.adoc:53]: Why not? One benefit I see is that "get rid of all this stuff" operations such as `git clean -xfd` will also get rid of the switch.

File: Changes [hunk 6]
```diff
@@ -582,6 +582,9 @@ OCaml 4.08.0
 * GPR#8533: Remove some unused configure tests
   (Stephen Dolan, review by David Allsopp and Sébastien Hinderer)
 
+- GPR#2207,#8604: Add opam files to allow pinning
+  (Leo White, Greta Yorsh)
```
Comment (gasche) [inline Changes]: @gretay-js please include Gabriel Radanne as a reviewer and I'll merge. Thanks both (and Leo) for the work!

**Comment by gretay-js** [comment]
Thank you!

**Comment by gasche** [comment]
I just merged, and this went into the 4.08 branch. If I understand correctly, I should update the version to 4.09 to cherry-pick in trunk, but this also requires adding a 4.09 version in opam-repository or using ocaml-beta-repository in the meantime.

**Comment by gretay-js** [comment]
Yes, as far as I understand, but I am not sure about the procedure for updating
opam repositories of ocaml vs release, so I haven't submitted any pull
requests for updating
https://github.com/ocaml/opam-repository/tree/master/packages/ocaml.
It's just one file which I attached to this PR.

**Comment by gasche** [comment]
I cherry-picked in trunk, but I didn't do anything on the opam-repository side. Anyone willing to prepare a change there is welcome, otherwise I'll get to it at some (unknown) later point.

**Comment by dra27** [comment]
Note that the ocaml-variants package on trunk doesn't want updating at _release_ it wants updating at _branch_ time cf https://github.com/ocaml/ocaml/pull/2207#issuecomment-448594122.

It's always OK to have `ocaml` packages in opam-repository for unreleased versions - at this moment, `ocaml.4.08.0` and `ocaml.4.09.0` could (and should) both be there. The key thing at the moment is that there must never be an `ocaml-variants` package **in opam-repository** which has a version greater than the latest `ocaml-base-compiler` **in opam-repository**.

So, when we *branch* 4.09, `ocaml-variants.opam` on **trunk** should immediately be updated to be 4.10 and `ocaml.4.10.0` should be merged to opam-repository.

**Comment by avsm** [comment]
https://github.com/ocaml/opam-repository/pull/13964 is now merged

**Comment by lpw25** [comment]
@dra27 Could we maybe revisit the decision to not include an `ocaml.opam` file? At the moment, pinning the compiler seems to work approximately none of the time because the opam-repository is not updated when the VERSION file is changed in a branch. For example, both `trunk` and `4.08` branches are currently on versions that I don't think have ocaml opam files in the repository -- and in the case of the `4.08` branch it is not even clear that the version should be added to the repository.

Is there any actual harm in keeping a local copy in the distribution? Maybe with a script to update the versions in it based on the VERSION file so it is easy to fix when things are changed (or even to check in CI that these things are in sync).

**Comment by dra27** [comment]
@lpw25 - sorry for the delay ACK'ing this. In the intervening time, https://github.com/ocaml/opam-repository/pull/14519 has been merged which should allow pinning on 4.08. The 4.10 package has been in opam repository for 3 months when https://github.com/ocaml/opam-repository/pull/14065 was merged. What's the error you were seeing with `trunk` pinning?

The issue seems to be more with the release checklist (or at least the actions which take place) - there should be an opam-repository PR to add the new ocaml package version for `trunk` when a release branch is cut and the pull request from final release candidate to release should also include the ocaml package version for the next point release i.e.

- 4.10 should be branched *after* a PR to opam-repository with `ocaml.4.11.0` has been merged
- The opam-repository PR releasing 4.09.0 should include `ocaml.4.09.1`

@gasche / @Octachron - where's appropriate in the release-checklist to note that?

**Comment by lpw25** [comment]
But do you actually gain anything from doing that rather than just including an opam file in the distribution? Unless there is a concrete benefit from this approach then it seems to just make things more fragile and generate extra work for people.

**Comment by dra27** [comment]
It doesn't generate work - those files all have to exist at some point in opam-repository, so it's just rescheduling work.

You do gain something when annoying Windows programmers push changes to the central `ocaml` packages and you then have `ocaml.4.10.0` in ocaml/opam-repository out-of-sync with the one in ocaml/OCaml when it really does generate more work. I'm not sure how code duplication can ever be more fragile than not code duplication!

**Comment by lpw25** [comment]
> I'm not sure how code duplication can ever be more fragile than not code duplication!

When the code is not part of the same repo and can therefore get out of sync. In that case you can make the code less fragile by "vendoring" it into your repo, duplicating the code but removing the potential for a mismatch.

**Comment by dra27** [comment]
`ocaml.opam` will systemically be out-of-sync - the file you'd need to put in the OCaml repository would need a `version` header. It's also not vendored, given that it retains the same name and therefore conflicts in surprising ways with any *downstream* update. I acknowledge it's weird, given the name, but the `ocaml` package has no philosophical or technical business being in the OCaml repo - it's part of opam (which also means it'd be vendoring a *downstream* file, not a dependency).

Most importantly, it also doesn't work. Embedding `ocaml.opam` in this repo works for `opam install .` but it doesn't work for `opam pin add`, where having it where it belongs in opam-repository does work. As an aside, `opam install .` is what you'd do to work on your branch, `opam pin add ocaml-variants git+https://github.com/lpw25/ocaml/...` is what you'd expect someone else to do to try out your changes and is also the rather poorly advertised way of testing out an OCaml PR in opam 2.

`trunk` should have been pinning correctly for the last 3 months (have you been seeing an error there, as I asked?) and 4.08.1 pinning has been fixed in the last few days. So let's fix the release procedure so that those files are never missing and fix it properly, rather than somewhat lazily adding a partially working bandaid into this repo.

**Comment by lpw25** [comment]
> ocaml.opam will systemically be out-of-sync

It might get out-of-sync but it is easy to put back in sync without the need for cooperation from the opam repository maintainers.

> but it doesn't work for opam pin add

This is at last a non-aesthetic reason to not put the file in the repo. Although I would appreciate an explanation of why and in what way it doesn't work, as it seems like that should work just fine.

> So let's fix the release procedure so that those files are never missing and fix it properly, rather than somewhat lazily adding a partially working bandaid into this repo.

It's got nothing to do with lazyness. Someone needed to opam pin their compiler based on 4.08 the other day and proceeded to get a series of completely unhelpful error messages and lost a couple of hours of productive time. That is somewhat frustrating and would have been avoided if the file was just in the repository to begin with.

**Comment by dra27** [comment]
By systemically out-of-sync, I'm meaning that `ocaml.opam` in the repo would *never* be the same as the one in opam-repository. Worse, if the one in the repo omits the `version` field (e.g. through changes from downstream being just copied in, or the ocaml.opam from a release being accidentally copied in) then you can get an even stranger error message about version numbers from the solver, since opam will assign the greatest version of ocaml it knows about to the pin.

When you add a pin, you are pinning the `ocaml-variants` package, so opam has no business scanning other opam files in the repo (so it won't pick up the `ocaml.opam` package unless you explicitly `opam pin add ocaml git+https://github.com/lpw25/ocaml/...` as well). Conversely, `opam install .` is an instruction to install every opam package found in the current directory, so it does pick up both `ocaml-variants.opam` and `ocaml.opam`.

In addition to fixing the release checklist, it sounds as though it would be worth updating `HACKING.adoc` with notes on the error message, just in case it happens again.

**Comment by lpw25** [comment]
> By systemically out-of-sync, I'm meaning that ocaml.opam in the repo would never be the same as the one in opam-repository.

Do you just mean because of the version field, or is there some other stuff that we can't/shouldn't put in the local version?

> Worse, if the one in the repo omits the version field...

Well that's pretty similar to saying "if someone writes a broken ocaml.opam file". You should always put a version in your pinning opam files.

> When you add a pin, you are pinning the ocaml-variants package, so opam has no business scanning other opam files in the repo (so it won't pick up the ocaml.opam package unless you explicitly opam pin add ocaml git+https://github.com/lpw25/ocaml/... as well).

That just seems like the expected behaviour. It means that if the opam file in the opam repository is working then you can just do `opam pin add ocaml-variants ...` and if it happens to be not up-to-date a simple `opam pin add ocaml ...` will get things working. And you can just always do both pins and not worry about the state of the opam repository.

**Comment by lpw25** [comment]
As another possibility, what about *just* having an `ocaml.opam` in the OCaml repository? Instead of pinning ocaml-variants and then relying on an `ocaml` package that depends on `ocaml-variants`, why not just pin an `ocaml` package that installs the compiler directly?

I know we can't do that in the main opam repository, but for pinning it seems like it would probably work fine. Then you could actually do `opam pin ocaml URL` -- which is what people would expect to be able to do to pin the version of OCaml that they are using.

**Comment by dra27** [comment]
That should work, yes - I think we might need a dummy `ocaml-variants` package in `opam-repository` to ensure that one of `ocaml-system`, `ocaml-base-compiler` and `ocaml-variants` is installed (although I haven't checked whether any package, internal or otherwise, actually references those packages directly).

There would be no relationship between `ocaml.opam` here and in opam-repository, so that would seem to allay my concerns about synchronising them. I'll try to do some tests on it this week - I haven't yet convinced myself what would want to happen to `ocaml-variants` pinning.

**Comment by gasche** [comment]
Could someone list here the steps to be performed, from the `.opam` files in the repository, to install a fresh switch with the current state of trunk? I tried to do it again yesterday (from the 4.09 branch I think) and failed.

**Comment by dra27** [comment]
It should be:

```sh
opam switch create . --empty
opam install .
```

(or `HACKING.adoc` needs an update!)

**Pull request merged**